src/grammar.rs

```diff
@@ -685,6 +685,137 @@ mod tests {
+        <char_null> ::= <char>
+        <char> ::= 'a'
+        <special-char> ::= <char_null> | <whitespace>
+        <whitespace> ::= ' '
+        <ws> ::= ' ' | ' ' <ws>
+        ";
+        assert!(left_recursive
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        let right_recursive = left_recursive.replace(
+            // rewrite production from left- to right- recursive
+            "<string> ::= <char_null> | <string> <char_null>",
+            "<string> ::= <char_null> | <char_null> <string>",
+        );
+        assert!(
+            right_recursive
+                .parse::<Grammar>()
+                .unwrap()
+                .parse_input(input)
+                .next()
+                .is_some(),
+            "right-recursive grammar failed to parse: {input}"
+        );
+    }
+
+    #[test]
+    fn shared_nullable_nonterminal() {
+        let fails: &str = "
+        <disjunction> ::= <predicate> | <disjunction> <or> <predicate>
+        <predicate> ::= <char_null_one> | <special-string> '.'
+
+        <char_null_one> ::= <char_null_two>
+        <char_null_two> ::= <char_null_three>
+        <char_null_three> ::= <char>
+
+        <or> ::= <ws> 'or' <ws>
+        <ws> ::= <whitespace> | ' ' <ws>
+        <whitespace> ::= ' '
+
+        <special-string> ::= <special-char> | <special-char> <special-string>
+        <special-char> ::= <char> | <whitespace>
+        <char> ::= 'a'
+        ";
+
+        let input = "a or a";
+
+        let passes_1 = fails.replace(
+            // skip nullable production <char_null_two>
+            "<char_null_one> ::= <char_null_two>",
+            "<char_null_one> ::= <char_null_three>",
+        );
+        assert!(passes_1
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        let passes_2 = fails.replace(
+            // replace <whitespace> with its terminal ' '
+            "<ws> ::= <whitespace> | ' ' <ws>",
+            "<ws> ::= ' ' | ' ' <ws>",
+        );
+        assert!(passes_2
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        let passes_3 = fails.replace(
+            // again, replace <whitespace> with its terminal ' '
+            "<special-char> ::= <char> | <whitespace>",
+            "<special-char> ::= <char> | ' '",
+        );
+        assert!(passes_3
+            .parse::<Grammar>()
+            .unwrap()
+            .parse_input(input)
+            .next()
+            .is_some());
+
+        assert!(
+            fails
+                .parse::<Grammar>()
+                .unwrap()
+                .parse_input(input)
+                .next()
+                .is_some(),
+            "all grammars except last one parsed: {input}"
+        );
+    }
```

Comment:
@amascolo I believe I have included your two reproduction cases verbatim here! If you see anything amiss, please let me know. I might end up renaming some of the variables (like `fail`) now that they are passing, but wanted to leave them untouched until you were done experimenting

Reply:
Hey, just saw this PR now – I'm currently travelling – will gladly experiment next week and let you know 🙂

Reply:
This is a definitely a step forward.

I have tested this branch locally –  it fixes the cases I had raised, without introducing any obvious regressions.

However, I am still seeing issues with nullable productions that fail to match: https://github.com/shnewto/bnf/issues/117#issuecomment-1377411514

On the larger grammar I'm writing, also still seeing cases where changing productions from left- to right- recursive fails to match – but it's tricky to investigate until the case above is fixed, so I might wait until that's resolved.

Reply:
@CrockAgile any chance of merging this PR and releasing it as `0.4.4` please?

It's definitely an improvement over `0.4.3` as it fixes the more noticeable edge cases that are failing.

Then you can close #118, while leaving #117 open for fixing the remaining cases – which are more involved and would probably warrant a separate PR anyway.